portalnet/src/metrics.rs

```diff
@@ -14,79 +18,99 @@ use crate::types::messages::ProtocolId;
+            // in testing situations. In regular usage, it should be reported as an error:
+            error!("Failed to register prometheus messaging metrics with default registry, creating new");
+
+            let custom_registry = Registry::new_custom(None, None)
+                .expect("Prometheus docs don't explain when it might fail to create a custom registry, so... hopefully never");
+            register_int_counter_vec_with_registry!(message_count_options, message_count_labels, custom_registry)
+                .expect("a gauge can always be added to a new custom registry, without conflict")
+        });
+
+        Self { message_count }
     }
 
-    pub fn report_inbound_find_nodes(&self) {
-        self.inbound_find_nodes.inc();
+    /// Returns the value of the given metric with the specified labels.
+    pub fn value_by_labels(&self, metric_id: MetricId, labels: &[&str]) -> u64 {
+        let metric = match metric_id {
+            MetricId::MessageCount => &self.message_count,
+        };
+        metric.with_label_values(labels).get()
     }
 
-    pub fn report_inbound_find_content(&self) {
-        self.inbound_find_content.inc();
+    pub fn report_outbound_request(&self, protocol: &ProtocolId, request: &Request) {
+        let message = Message::from(request.clone());
+        self.increment_message_count(protocol, "sent", message);
     }
 
-    pub fn report_inbound_offer(&self) {
-        self.inbound_offer.inc();
+    pub fn report_inbound_request(&self, protocol: &ProtocolId, request: &Request) {
+        let message = Message::from(request.clone());
+        self.increment_message_count(protocol, "received", message);
     }
 
-    pub fn report_outbound_ping(&self) {
-        self.outbound_ping.inc();
+    pub fn report_outbound_response(&self, protocol: &ProtocolId, response: &Response) {
+        let message = Message::from(response.clone());
+        self.increment_message_count(protocol, "sent", message);
     }
 
-    pub fn report_outbound_find_nodes(&self) {
-        self.outbound_find_nodes.inc();
+    pub fn report_inbound_response(&self, protocol: &ProtocolId, response: &Response) {
+        let message = Message::from(response.clone());
+        self.increment_message_count(protocol, "received", message);
     }
 
-    pub fn report_outbound_find_content(&self) {
-        self.outbound_find_content.inc();
+    fn increment_message_count(&self, protocol: &ProtocolId, direction: &str, message: Message) {
+        let protocol_label = Self::protocol_to_label(protocol);
+        let message_type_label = Self::message_to_label(message);
+        let labels = [protocol_label, direction, message_type_label];
+        self.message_count.with_label_values(&labels).inc();
     }
 
-    pub fn report_outbound_offer(&self) {
-        self.outbound_offer.inc();
+    fn message_to_label(message: Message) -> &'static str {
```

Comment:
to avoid cloning each message to increment the corresponding metric, we could implement conversions from the `&Request` and `&Response` types